Os 6 key api

LeonardoJPerez opened the pull request:
Refactored and updated unit tests.
Decoupled KeyHandler from dbContext via new interface.
Added initial Dockerfile.

Commit:
Debug configuration.

Commit:
Updated model annotations.

Added constructor to JSONResponse.

Commit:
Implemented Get and Insert methods.

Moved configuration values to be read from env.

Commit:
Wired up db interaction from within endpoint handlers.

Insert and Get so far.

Commit:
Implemented delete and update db methods.

Updated Key model with proper PK.

Commit:
Clean up.

Commit:
Added Dockerfile.

Commit:
Added dbcontext dependency to KeyHandlers.

Created new interface for dbContext for testability.
Changed Key pk type to Int.
Stubbedout dbcontext unit tests.

Commit:
Added mockDbContext struck for tests.

Updated KeyHandlers tests to use new mocked dependency.

Commit:
Removed test stub.

kcoleman731 commented on db/dbContext.go:
```diff
@@ -1,4 +1,128 @@
+		log.Printf("%s %s", time.Since(event.StartTime), query)
+	})
+
+	// err := dbc.CreateSchema()
```
Commented code. Either uncomment or ditch in a pr.

LeonardoJPerez replied:
Good catch. Code removed.

kcoleman731 commented on db/dbContext.go:
```diff
@@ -1,4 +1,128 @@
+	return nil
+}
+
+func (dbc *dbContext) GetModel(keyID int) models.Key {
```
for these model specific query functions, lets put them in the model. Also be descriptive with the naming. so `db.KeyWithID()`

kcoleman731 replied:
Docs on all exported functions.

LeonardoJPerez replied:
Done!

kcoleman731 commented on db/dbContext.go:
```diff
@@ -1,4 +1,128 @@
+	return aKey
+}
+
+func (dbc *dbContext) GetModels(userID string) []models.Key {
```
Same thing here. `db.KeysForUserID`

LeonardoJPerez replied:
Done!

kcoleman731 commented on db/dbContextInterface.go:
```diff
@@ -0,0 +1,13 @@
+import "2FAServer/models"
+
+// DbContextInterface for DB access.
+type DbContextInterface interface {
```
If you have a dbContext class, its interface should be declared in the same file.

LeonardoJPerez replied:
Done!

kcoleman731 commented on handler/keyHandler_test.go:
```diff
@@ -1,77 +1,105 @@
-	deleteKeyJSONResponse = `{"timestamp":` + strconv.Itoa(int(time.Now().Unix())) + `,"message":"` + testUserID + `"}`
-	testUserID            = "1e3243566776998723t3reververv"
+	h       = handler.KeyHandler{DbContext: new(db.MockDbContext)}
+	testKey = models.Key{
```
we should bring in a factory package, so we don't have to hardcode values. https://github.com/icrowley/fake

LeonardoJPerez replied:
Done! Added fake and updated unit tests.

Commit:
PR comments and additional documentation.

Commit:
Added fake to unit tests.

LeonardoJPerez commented:
Thanks @kcoleman731 for the feedback.

Commit:
Added env variables to code and dockerfile.

Added .dockerignore.

Commit:
Refactored code to decouple Key model from DbContext.

Created new Store layer and interfaces.
Wire up handler to new store.
Fixed db abstraction.
Updated unit tests.

tayhalla commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+ENV PG_DATABASE="2FactorAuth"
+ENV PG_HOSTNAME="127.0.0.1"
+ENV PG_PORT="32768"
+ENV GOPATH="/go"
```
fwiw - im suprised youre setting this explictly - isn't it set from the base image?

LeonardoJPerez replied:
Yes it is. I was just trying to match my env to the image, but yeah, this statement is not really necessary. Ill take it out.

Commit:
Fixed type issue when receiving model in DbContext.

tayhalla commented on db/mockDbContext.go:
```diff
@@ -0,0 +1,110 @@
+
+// UpdateKey updates a Key records's key value.
+func (dbc *MockDbContext) UpdateModel(model interface{}) bool {
+	method := reflect.ValueOf(model).MethodByName("ObjectID")
```
ok - so if we're expecting this `model` to have a method called ObjectID, why are we not defining that in an interface instead of accepting a blank interface. ie `interface{}`

LeonardoJPerez replied:
I was getting a runtime error (within PG) when the func expected a model.Persistable. I think I was passing the data the wrong way. I definitely rather have a func that expects a specific interface not the generic one. I'll change it and if I still hit a wall Ill let you know.

tayhalla replied:
ok - sounds good. Ya, obviously refection is cool, but let's always strive to be as strongly typed as possible

kcoleman731 commented on store/keyStore.go:
```diff
@@ -0,0 +1,72 @@
+package store
+
+import (
```
need test coverage on this guy.

kcoleman731 replied:
otherwise looks good.

LeonardoJPerez replied:
Definetly 😄 . Thanks!!!

Commit:
Updated underlying db context struct for abstraction.

Commit:
Added keyStore tests.
